lambdas/data-migration1/src/rules.ts

```diff
@@ -0,0 +1,141 @@
+import Knex from 'knex';
+
+import DynamoDbSearchQueue from '@cumulus/aws-client/DynamoDbSearchQueue';
+import { PostgresRuleRecord } from '@cumulus/db';
+import { envUtils } from '@cumulus/common';
+import { RecordDoesNotExist } from '@cumulus/errors';
+import Logger from '@cumulus/logger';
+
+import { RecordAlreadyMigrated } from './errors';
+import { MigrationSummary } from './types';
+
+const logger = new Logger({ sender: '@cumulus/data-migration/rules' });
+const Manager = require('@cumulus/api/models/base');
+const schemas = require('@cumulus/api/models/schemas');
+
+/**
+ *
+ * Retrieve cumulus_id from table using name and version.
+ *
+ * @param {object} whereClause - where clause for query
+ * @param {'providers' | 'collections'} table - Name of table
+ * @param {Knex} knex - Knex client for writing to RDS database
+ * @returns {Promise<number>} - Cumulus ID for record
+ * @throws {RecordDoesNotExist} if record cannot be found
+ */
+export const getCumulusId = async<T>(
+  whereClause : object,
+  table: 'providers' | 'collections',
+  knex: Knex
+): Promise<any> => {
+  const record = await knex.select('cumulus_id')
+    .from<T>(table)
+    .where(whereClause)
+    .first();
+
+  if (record === undefined) {
+    throw new RecordDoesNotExist(`Record in ${table} with identifiers ${whereClause} does not exist.`);
+  }
+  return record;
```

Comment:
I don't understand why we changed this function to return the record instead of record.cumulus_id

Reply:
I was running into this error when I tried to return `record.cumulus_id`
```
Property 'cumulus_id' does not exist on type 'UnknownToAny<(AnyToUnknown<T> extends {} ? Boxed<"cumulus_id"> extends Boxed<keyof T> ? SafePick<T, "cumulus_id" & keyof T> : any : unknown) | undefined>'.
```

I attempted to define the function to have `<T extends { cumulus_id: number }>` and also extending an interface I wrote that has the same property to no avail. 

Do you have any insight? I would also prefer to just return `record.cumulus_id`

Reply:
Yeah, it comes down to the function not being aware of `cumulus_id` existing on the record now that we've introduced a generic type.  Let me take a look at it

Reply:
What you tried is supposed to work: https://www.typescriptlang.org/docs/handbook/generics.html#generic-constraints

So the reason it doesn't may have more to do with Knex. I'll look into it.

Reply:
This is very tricky. The problem is that we're still applying the generic type to Knex here:

`.from<T>(table)`

And `T` is a generic type, so it doesn't specify that `cumulus_id` must be a part of it, hence the error. 

For now, I'd just drop the generic and do this:

```
interface CumulusRecord {
  cumulus_id: number
}

export const getCumulusId = async (
  whereClause : object,
  table: 'providers' | 'collections',
  knex: Knex
): Promise<number> => {
  const record = await knex.select('cumulus_id')
    .from<CumulusRecord>(table)
    .where(whereClause)
    .first();

  if (record === undefined) {
    throw new RecordDoesNotExist(`Record in ${table} with identifiers ${whereClause} does not exist.`);
  }
  return record.cumulus_id;
};
```

However, this is less than ideal, because someone could in theory run this function against a table where `cumulus_id` does not exist, and it wouldn't fail TS compilation, but would fail at runtime. You can try to look into a solution if you want, but I would timebox it to 30 minutes.

Reply:
Okay, I will look into this but timebox to 30mins. Thanks for looking into this further! I think it is not great that it would fail at runtime so I'm going to try to see a way to prevent that.

Reply:
See #2006 for a proposed refactor